Adapter and quality trimming for IlluminaBasecallsToFastq

jacarey opened the pull request:
### Description

This adds adapter and quality trimming to the fastq outputs.

Questions:

1) which should occur first quality or adapter trimming? Quality trimming is less computationally expensive so it may be _slightly_ more efficient to do that first unless there is a reason to do otherwise.
2) We are modifying the default behavior to include adapter trimming (similar to IBCTS) should we instead default to no adapter trimming?
----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [ ] Added or modified tests to cover changes and any new functionality
- [ ] Edited the README / documentation (if applicable)
- [ ] All tests passing on Travis

#### Review
- [ ] Final thumbs-up from reviewer
- [ ] Rebase, squash and reword as applicable

For more detailed guidelines, see https://github.com/broadinstitute/picard/wiki/Guidelines-for-pull-requests

Commit:
Add adapter and quality trimming to basecalls to fastq.

tfenne commented on src/main/java/picard/illumina/IlluminaBasecallsToFastq.java:
```diff
@@ -486,17 +529,48 @@ public void write(final ClusterData rec) {
+                byte[] quals = read.getQualities();
+
+                if (trimmingQuality != null) {
+                    int index = TrimmingUtil.findQualityTrimPoint(quals, trimmingQuality);
+                    // Don't write zero length reads if the entire read is trimmed
+                    if (index == 0) {
+                        return;
+                    }
+                    quals = Arrays.copyOfRange(quals, 0, index);
+                    bases = Arrays.copyOfRange(bases, 0, index);
+                }
```
So I don't think we can do the `if (index == 0)` test here.  If we have PE reads, or even say reads with extracted barcodes or UMIs being written to separate fastqs, then we could skip e.g. R1 but include R2 and cause record/line numbers to go out of sync.  

Also is this applied to all reads regardless of type?  I think we should not be adapter or quality trimming anything but template reads.

I think I would suggest moving the trimming-related code so that it's called a) only on template reads and b) directly from write(ClusterData).  That way we can e.g. skip over clusters where all reads are trimmed to 0bp, and otherwise do something more sensible.  Perhaps there's a `MIN_TRIMMED_LENGTH` or similar parameter, that controls this?  I.e. if that's set to 20bp and quality trimming says remove all of it (but not it's mate), then we only trim to 20bp?

tfenne replied:
I think the adapter marking would also be better too since you should in theory find the adapter at the same base in each read and that boosts confidence.

tfenne commented on src/main/java/picard/illumina/IlluminaBasecallsToFastq.java:
```diff
@@ -486,17 +529,48 @@ public void write(final ClusterData rec) {
+                }
+
+                if (adapterMarker != null) {
+                    AdapterPair adapterPair = adapterMarker.findAdapterPairForSingleRead(bases, templateIndex);
+                    if (adapterPair != null) {
+                        int index = bases.length - 1;
+                        if (templateIndex == 1) {
+                            index = ClippingUtility.findIndexOfClipSequence(
+                                    bases, adapterPair.get3PrimeAdapterBytes(), ClippingUtility.MIN_MATCH_BASES, ClippingUtility.MAX_ERROR_RATE);
+                        } else if (templateIndex == 2) {
+                            index = ClippingUtility.findIndexOfClipSequence(
+                                    bases, adapterPair.get5PrimeAdapterBytesInReadOrder(), ClippingUtility.MIN_MATCH_BASES, ClippingUtility.MAX_ERROR_RATE);
+                        }
```
I had to go look in ClippingUtility and it's also like this, which is crazy!  So we search the read once in order to figure out _which_ adapter pair it is, and then if we got a match, we go back and do it _again_ to re-compute where it is?  Ugh.

Could we maybe add methods to AdapterMarker/ClippingUtiliity that return a tuple of `<AdapterPair, Integer>` to give you both at the same time and avoid re-matching?

jacarey commented on src/main/java/picard/illumina/IlluminaBasecallsToFastq.java:
```diff
@@ -190,6 +199,13 @@
+    public Integer TRIMMING_QUALITY = null;
+
+    @Argument(doc = "The minimum length for a trimmed read. " +
+            "If trimming would create a smaller read trim to this length instead", optional = true)
```
```suggestion
            "If trimming would create a smaller read, then trim to this length instead", optional = true)
```

Commit:
Merge branch 'master' into jc_adapter_quality_trimming

Commit:
Only read the cbcl header once the tile is about to be cached and then close the streams. (#1686)

* Only read header once the tile is about to be cached and then close the cbcl streams.
* Fix header read for check illumina dir.
* Close provider after we read all clusters.

Commit:
[bugfix] Pass MAX_RECORDS_IN_RAM to BasecallsConverter (#1687)

Commit:
Jc fix small data race condition (#1690)

* [bugfix] Pass MAX_RECORDS_IN_RAM to BasecallsConverter
* Use a different thread pool executor for each tile to ensure no race conditions can occur

Commit:
bugfix and nio access in Fingerprinting code (no need to panic) (#1694)

* - Convert CalculateFingerprintMetrics, ExtractFingerprint, and ExtractContamination to use NIO and thus be able to to their work on a non-localized input file.
- Fixed a bug in the AD output of extractFingerprint and CalculateContamination (Ref and Alt were swapped...)
- Renamed PL variable to be GL since that's what is actually was.

* -removed unneeded mathUtil method.

Commit:
fix: underflow in HaplotypeProbabilitiesFromContaminatorSequence (#1697)

* fix: underflow in HaplotypeProbabilitiesFromContaminatorSequence, which affects both IdentifyContamination and ExtractFingerprint in deep coverage. This was leading to non-informative sites in RNA and similar sequencing data.

Commit:
Merge branch 'master' into jc_adapter_quality_trimming

mollysacks committed:
Fix bug in CollectSamErrorMetrics (#1618)

* first

* place-holder

* testing

* testing

* delete-logfiles

* rm-ds-store

* Update CollectSamErrorMetrics.java

* add-test-case

Co-authored-by: Molly Sacks <msacks@broadinstitute.org>

mollysacks committed:
HaplotypeMap bug fix (#1702)

* use-basesEqual

* testing

Co-authored-by: Molly Sacks <msacks@broadinstitute.org>

gbggrant committed:
Fix typo in usage (#1703)

Commit:
Add on the fly demux and speed up barcode matching. (#1649)

* [bugfix] Pass MAX_RECORDS_IN_RAM to BasecallsConverter
* Add on the fly demux and speed up barcode matching

Commit:
Merge branch 'master' into jc_adapter_quality_trimming

Commit:
Apply suggestions from code review

Co-authored-by: Jessica Way <jway@broadinstitute.org>
Co-authored-by: George Grant <ggrant@broadinstitute.org>

Commit:
Review changes.

Commit:
Fix imports

Commit:
Make cram valid input to remaining CLPs  (#1622)

 Enable cram in more CLPs

fixes: #433, #431, #361

rickymagner committed:
Added some more skew-resistent metrics for read length (#1706)

Co-authored-by: Ricky Magner <rmagner@wm349-9a3.broadinstitute.org>

rickymagner committed:
Cleaned up some variable names and order in the alignment metrics code. (#1707)

Co-authored-by: Ricky Magner <rmagner@wm349-9a3.broadinstitute.org>

kachulis committed:
Adjust handling of missing data situations in checkFingerprints (#1700)

Commit:
Add adapter and quality trimming combined test.

Commit:
Merge branch 'master' into jc_adapter_quality_trimming
